source_pubsub/client.go

```diff
@@ -127,18 +130,22 @@ func (c *PubSubClient) HasNext(ctx context.Context) bool {
 
 // Next returns the next record from the buffer.
 func (c *PubSubClient) Next(ctx context.Context) (sdk.Record, error) {
+	sdk.Logger(ctx).Debug().Msgf("Next - number of records in buffer %d", len(c.buffer))
 	select {
-	case r := <-c.buffer:
-		sdk.Logger(ctx).Debug().Msgf("next record - %v", r)
-		return r, nil
-	case <-c.tomb.Dead():
+	case <-c.tomb.Dying():
 		err := c.tomb.Err()
-		sdk.Logger(ctx).Debug().Msgf("pubsub client tombstone.Dead(), err=%s", err)
-		return sdk.Record{}, fmt.Errorf("pubsub client tombstone.Dead(), err=%s", err)
+		sdk.Logger(ctx).Debug().Msgf("pubsub client tombstone.Dying(), err=%s", err)
+		return sdk.Record{}, fmt.Errorf("pubsub client tombstone.Dying(), err=%s", err)
 	case <-ctx.Done():
 		err := ctx.Err()
 		sdk.Logger(ctx).Debug().Msgf("pubsub client context.Done(), err=%s", err)
 		return sdk.Record{}, fmt.Errorf("pubsub client context.Done(), err=%s", err)
+	case r, ok := <-c.buffer:
+		if !ok {
+			return sdk.Record{}, fmt.Errorf("pubsub client, buffer is closed")
```

Comment:
check for the tomb error here.

Reply:
If `!ok` is true, then the tomb is truly dead, in which case we want to see if there were any errors and return them. So something like this should be added in here.

```
if !ok {
  if err := c.tomb.Err(); err != nil {
      return sdk.Record{}, fmt.Errorf("tomb exited: %w", err)
  }
  return sdk.Record{}, fmt.Errorf("end of records, buffer closed")
}
```